improver_tests/calibration/samos_calibration/test_TrainGAMsForSAMOS.py

```diff
@@ -0,0 +1,271 @@
+
+from copy import deepcopy
+
+import numpy as np
+import pytest
+from iris.coords import CellMethod
+from iris.cube import CubeList
+
+from improver.calibration.samos_calibration import TrainGAMsForSAMOS
+from improver_tests.calibration.samos_calibration.helper_functions import (
+    create_cubes_for_gam_fitting,
+    create_simple_cube,
+)
+
+np.random.seed(1)
+
+
+@pytest.fixture
+def model_specification():
+    """Fixture for creating a model specification as used in SAMOS plugins."""
+    return [["linear", [0], {}], ["linear", [1], {}]]
+
+
+@pytest.mark.parametrize(
+    "kwargs",
+    [
+        {
+            "model_specification": [["linear", [0], {}]]
+        },  # Define a model specification but leave all other inputs as default.
+        {
+            "model_specification": [["linear", [0], {}]],
+            "max_iter": 200,
+            "tol": 0.1,
+        },  # Check that inputs related to model fitting are initialised correctly.
+        {
+            "model_specification": [["linear", [0], {}]],
+            "distribution": "gamma",
+            "link": "inverse",
+            "fit_intercept": False,
+        },  # Check that inputs related to the model design are initialised correctly.
+    ],
+)
+def test__init__(kwargs):
+    """Test that the class initializes variables correctly."""
+    # Skip test if pyGAM not available.
+    pytest.importorskip("pygam")
+
+    # Define the default, then update with any differently specified inputs.
+    expected = {
+        "model_specification": None,
+        "max_iter": 100,
+        "tol": 0.0001,
+        "distribution": "normal",
+        "link": "identity",
+        "fit_intercept": True,
+    }
+    expected.update(kwargs)
+    result = TrainGAMsForSAMOS(**kwargs)
+
+    for key in kwargs.keys():
+        assert getattr(result, key) == kwargs[key]
+
+
+@pytest.mark.parametrize("forecast_type", ["gridded", "spot"])
+@pytest.mark.parametrize("n_realizations,n_times", [[2, 1], [2, 2], [1, 2]])
+@pytest.mark.parametrize("include_blend_time", [False, True])
+def test_calculate_cube_statistics(
+    forecast_type,
+    n_realizations,
+    n_times,
+    include_blend_time,
+):
+    """Test that this method correctly calculates the mean and standard deviation of
+    the input cube."""
+    create_cube_kwargs = {
+        "forecast_type": forecast_type,
+        "n_spatial_points": 2,
+        "n_realizations": n_realizations,
+        "n_times": n_times,
+        "fill_value": 305.0,
+    }
+    expected_cube_kwargs = {
+        "forecast_type": forecast_type,
+        "n_spatial_points": 2,
+        "n_realizations": 1,
+        "n_times": n_times,
+    }
+
+    input_cube = create_simple_cube(**create_cube_kwargs)
+
+    # Create cubelist containing expected mean and standard deviations cubes.
+    expected_mean = create_simple_cube(fill_value=305, **expected_cube_kwargs)
+    expected_sd = create_simple_cube(fill_value=0, **expected_cube_kwargs)
```

Comment:
Given we need to be testing that the function is correctly calculating the mean over time periods it would be better to have different fill_values in each cube to confirm its correctly calculating a mean rather than just picking the fill value

Reply:
Good point - I've modified this test.